File: Algorithms/C++/00083-Remove Duplicates from Sorted List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

struct ListNode{
    int val;
    ListNode* next;
    ListNode* prev;
    
}

class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if(head == nullptr) return nullptr;
        ListNode* root = head;
        ListNode* prev = head;
        head = head->next;
        while(head != nullptr){
            if(head->val == prev->val){
                prev->next = head->next;
            }
            if(head->val != prev->val){
                prev = head;
            }
            head = head->next;
        }

        return root;
    }
};

/*
    1 -> 1 -> 2
    1 -> 2

*/

File: Algorithms/Python3/00083-Remove Duplicates from Sorted List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return head
        curr = head
        initVal = curr.val

        while curr:
            prev = curr
            while curr and curr.val == initVal:
                curr = curr.next
                if not curr:
                    prev.next = None
                    return head
            prev.next = curr
            initVal = curr.val

        return head
'''
    1) use extra memory (set)
    2) update next pointers in place

    iterate through list
    store currVal
    while node is equal to currVal
        skip over until new value is seen
    handle case for last number

    tc: O(n)
    sc: O(1)
'''